File: AlgorithmsPackPythonSolutions/264. Ugly Number II.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        minHeap = [1]
        heapq.heapify(minHeap)
        
        nth = 0
        mnValue = 0
        
        while minHeap and nth < n:
            nth += 1
            mnValue = heapq.heappop(minHeap)
            
            while minHeap and minHeap[0] == mnValue:
                heapq.heappop(minHeap)
            
            heapq.heappush(minHeap, 2 * mnValue)
            heapq.heappush(minHeap, 3 * mnValue)
            heapq.heappush(minHeap, 5 * mnValue)
        
        return mnValue

File: AlgorithmsPackCPP/264. Ugly Number II.cpp
class Solution {
    typedef long long ll;  //avoiding overflow
public:
    int nthUglyNumber(int n) {
        priority_queue<ll,vector<ll>,greater<ll>>minHeap;
        minHeap.push(1);
        
        int nth = 0;
        ll mnValue;
        
        while(!minHeap.empty() && nth < n){
            nth+=1;
            mnValue = minHeap.top();
            while(!minHeap.empty()&&minHeap.top()==mnValue)
                minHeap.pop();
            
            minHeap.push(2*mnValue);
            minHeap.push(3*mnValue);
            minHeap.push(5*mnValue);
        }
        return mnValue;
    }
};